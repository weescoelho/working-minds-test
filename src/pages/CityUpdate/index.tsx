import { Button, TextField } from '@material-ui/core';
import SaveIcon from '@material-ui/icons/Save';
import { Header } from 'components/Header';
import { DataContext } from 'contexts/DataContext';
import { Loading } from 'helpers/Loading';
import { useContext } from 'react';
import { FormEvent, useState } from 'react';
import { useHistory, useParams } from 'react-router-dom';
import styled from 'styled-components';

const Container = styled.div`
  max-width: 550px;
  margin: 0 auto;
  form {
    display: flex;
    flex-direction: column;
  }
  h2 {
    margin: 64px 0px 16px 0;
  }
  button {
    margin-top: 32px;
  }
`;

type StatePageParams = {
  id: string;
  name: string;
  stateId: string;
};

export function CityUpdate(): JSX.Element {
  const { updateCity, getData } = useContext(DataContext);
  const [newCityName, setNewCityName] = useState('');
  const [loading, setLoading] = useState(false);
  const params = useParams<StatePageParams>();
  const history = useHistory();

  const handleStateSubmit = async (event: FormEvent): Promise<void> => {
    event.preventDefault();
    setLoading(true);
    await updateCity(newCityName, params.id, params.stateId);
    await getData();
    setLoading(false);
    history.push(`/estado/${params.stateId}`);
    setNewCityName('');
  };

  return (
    <>
      {loading && <Loading />}
      <Header />
      <Container>
        <h2>Editar cidade: {params.name}</h2>
        <form onSubmit={handleStateSubmit}>
          <TextField
            id="outlined-helperText"
            label="Novo nome"
            helperText="Digite o nome da cidade"
            variant="outlined"
            onChange={({ target }) => setNewCityName(target.value)}
          />
          <Button
            variant="contained"
            type="submit"
            color="primary"
            startIcon={<SaveIcon />}
          >
            Alterar nome
          </Button>
        </form>
      </Container>
    </>
  );
}
