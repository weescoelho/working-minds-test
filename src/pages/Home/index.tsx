import React, { useState } from 'react';
import styled from 'styled-components';
import { useHistory } from 'react-router-dom';
import { api } from 'api/api';
import { Header } from 'components/Header';
import { City, States } from 'utils/types';
import { Button } from '@material-ui/core';
import { StateCard } from '../../components/StateCard';

const Container = styled.div`
  max-width: 900px;
  margin: 0 auto;
`;

const StateGrid = styled.div`
  display: grid;
  grid-template-columns: 1fr 1fr 1fr;
  column-gap: 16px;
  row-gap: 16px;
  margin: 64px 0;
`;

export function Home(): JSX.Element {
  const [data, setData] = useState<States>();
  const history = useHistory();

  React.useEffect(() => {
    const getData = async (): Promise<void> => {
      const stateData = await api.get('/states');
      setData(stateData.data);
    };
    getData();
  }, []);

  const handleCreateState = (): void => {
    history.push('/cadastrar/estado');
  };

  return (
    <>
      <Header />
      <Container>
        <StateGrid>
          {data &&
            data.map(state => (
              <StateCard key={state.id} id={state.id} stateName={state.name} />
            ))}
        </StateGrid>
        <Button variant="contained" color="primary" onClick={handleCreateState}>
          Adicionar estado
        </Button>
      </Container>
    </>
  );
}
