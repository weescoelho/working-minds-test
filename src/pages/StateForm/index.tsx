import { Button, TextField } from '@material-ui/core';
import { api } from 'api/api';
import { Header } from 'components/Header';
import { Loading } from 'helpers/Loading';
import { FormEvent, useState } from 'react';
import { useHistory } from 'react-router-dom';
import { useToasts } from 'react-toast-notifications';
import styled from 'styled-components';

const Container = styled.div`
  max-width: 550px;
  margin: 0 auto;
  form {
    display: flex;
    flex-direction: column;
  }
  h2 {
    margin: 64px 0px 16px 0;
  }
  button {
    margin-top: 32px;
  }
`;

export function StateForm(): JSX.Element {
  const [state, setState] = useState('');
  const [loading, setLoading] = useState(false);
  const { addToast } = useToasts();
  const history = useHistory();

  function handleStateSubmit(event: FormEvent): void {
    event.preventDefault();
    const createState = async (): Promise<any> => {
      try {
        await api.post('/states', {
          name: state,
        });
        addToast('Estado cadastrado com sucesso', {
          appearance: 'success',
          autoDismiss: true,
        });
        history.push('/');
      } catch {
        addToast('NÃ£o foi possivel cadastrar!', {
          appearance: 'error',
          autoDismiss: true,
        });
      }
    };
    createState();
    setState('');
  }

  return (
    <>
      {loading && <Loading />}
      <Header />
      <Container>
        <h2>Cadastrar Estado</h2>
        <form onSubmit={handleStateSubmit}>
          <TextField
            id="outlined-helperText"
            label="Nome da cidade"
            helperText="Digite o nome da cidade"
            variant="outlined"
            onChange={({ target }) => setState(target.value)}
          />
          <Button variant="contained" type="submit" color="primary">
            Cadastrar
          </Button>
        </form>
      </Container>
    </>
  );
}
