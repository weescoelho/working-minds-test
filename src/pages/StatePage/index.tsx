import { Button, TextField } from '@material-ui/core';
import SaveIcon from '@material-ui/icons/Save';
import DeleteIcon from '@material-ui/icons/Delete';
import { api } from 'api/api';
import { CityCard } from 'components/CityCard';
import { Header } from 'components/Header';
import { DataContext } from 'contexts/DataContext';
import { Loading } from 'helpers/Loading';
import { useContext } from 'react';
import { FormEvent, useEffect, useState } from 'react';
import { useHistory, useParams } from 'react-router-dom';
import { useToasts } from 'react-toast-notifications';
import { City, State } from 'utils/types';
import {
  Container,
  CreateNewCityArea,
  SectionHeader,
} from './StatePage.styled';

type StatePageParams = {
  id: string;
};

export function StatePage(): JSX.Element {
  const { deleteStateById, createNewCity } = useContext(DataContext);
  const [dataCities, setDataCities] = useState<Array<City>>();
  const [dataState, setDataState] = useState<State>();
  const [loading, setLoading] = useState(false);
  const params = useParams<StatePageParams>();
  const history = useHistory();
  const { addToast } = useToasts();
  const [newCity, setNewCity] = useState('');

  useEffect(() => {
    const getData = async (): Promise<void> => {
      setLoading(true);
      const cities = await api.get(`/cities?stateId=${params.id}`);
      const states = await api.get(`/states/${params.id}`);
      setDataCities(cities.data);
      setDataState(states.data);
      setLoading(false);
    };
    getData();
  }, [params.id]);

  const getDataOnUpdate = async (): Promise<void> => {
    setLoading(true);
    const cities = await api.get(`/cities?stateId=${params.id}`);
    const states = await api.get(`/states/${params.id}`);
    setDataCities(cities.data);
    setDataState(states.data);
    setLoading(false);
  };

  const handleNewCitySubmit = async (event: FormEvent): Promise<void> => {
    event.preventDefault();
    setLoading(true);
    await createNewCity(newCity, params.id);
    getDataOnUpdate();
    setLoading(false);
    setNewCity('');
  };

  const handleUpdateStateName = (): void => {
    history.push(`/estado/editar/${params.id}/${dataState?.name}`);
  };

  const handleDeleteState = async (): Promise<void> => {
    await deleteStateById(params.id);
    addToast('Estado excluido com sucesso!', {
      appearance: 'success',
      autoDismiss: true,
    });
    history.push('/');
  };

  const handleReturnPage = (): void => {
    history.push('/');
  };

  return (
    <>
      {loading && <Loading />}
      <Header />
      <Container>
        <SectionHeader>
          <div>
            <h2>{dataState?.name}</h2>
            <p>Cidades cadastradas: {dataCities?.length}</p>
          </div>
          <div>
            <Button
              variant="outlined"
              type="submit"
              color="primary"
              onClick={handleUpdateStateName}
            >
              Editar
            </Button>
            <Button
              variant="outlined"
              type="submit"
              color="secondary"
              onClick={handleDeleteState}
              startIcon={<DeleteIcon />}
            >
              Excluir estado
            </Button>
          </div>
        </SectionHeader>
        {dataCities &&
          dataCities.map(city => (
            <CityCard
              key={city.id}
              stateId={params.id}
              cityName={city.name}
              cityId={city.id}
              setDataCities={setDataCities}
            />
          ))}
        <CreateNewCityArea>
          <h3>Cadastrar nova cidade</h3>
          <form onSubmit={handleNewCitySubmit}>
            <TextField
              id="outlined-helperText"
              label="Nome da cidade"
              helperText="Digite o nome da cidade"
              variant="outlined"
              onChange={({ target }) => setNewCity(target.value)}
              defaultValue={newCity}
            />
            <Button
              variant="contained"
              type="submit"
              color="primary"
              startIcon={<SaveIcon />}
            >
              Cadastrar
            </Button>
          </form>
        </CreateNewCityArea>
        <Button variant="contained" color="primary" onClick={handleReturnPage}>
          Voltar
        </Button>
      </Container>
    </>
  );
}
