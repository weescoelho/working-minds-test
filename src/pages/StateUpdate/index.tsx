import { Button, TextField } from '@material-ui/core';
import SaveIcon from '@material-ui/icons/Save';
import { Header } from 'components/Header';
import { DataContext } from 'contexts/DataContext';
import { Loading } from 'helpers/Loading';
import { useContext } from 'react';
import { FormEvent, useState } from 'react';
import { useHistory, useParams } from 'react-router-dom';
import styled from 'styled-components';

const Container = styled.div`
  max-width: 550px;
  margin: 0 auto;
  form {
    display: flex;
    flex-direction: column;
  }
  h2 {
    margin: 64px 0px 16px 0;
  }
  button {
    margin-top: 32px;
  }
`;

type StatePageParams = {
  id: string;
  name: string;
};

export function StateUpdate(): JSX.Element {
  const { updateState, getData } = useContext(DataContext);
  const [newStateName, setNewStateName] = useState('');
  const [loading, setLoading] = useState(false);
  const params = useParams<StatePageParams>();
  const history = useHistory();

  const handleStateSubmit = async (event: FormEvent): Promise<void> => {
    event.preventDefault();
    setLoading(true);
    await updateState(newStateName, params.id);
    await getData();
    setLoading(false);
    history.push(`/estado/${params.id}`);
    setNewStateName('');
  };

  const handleReturnPage = (): void => {
    history.push(`/estado/${params.id}`);
  };

  return (
    <>
      {loading && <Loading />}
      <Header />
      <Container>
        <h2>Editar estado: {params.name}</h2>
        <form onSubmit={handleStateSubmit}>
          <TextField
            id="outlined-helperText"
            label="Novo nome"
            helperText="Digite o nome do estado"
            variant="outlined"
            onChange={({ target }) => setNewStateName(target.value)}
          />
          <Button
            variant="contained"
            type="submit"
            color="primary"
            startIcon={<SaveIcon />}
          >
            Alterar nome
          </Button>
        </form>
        <Button variant="contained" color="primary" onClick={handleReturnPage}>
          Voltar
        </Button>
      </Container>
    </>
  );
}
