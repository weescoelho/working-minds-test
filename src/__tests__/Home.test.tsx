import { create } from 'react-test-renderer';
import { Home } from '../pages/Home';

test('render Home page', () => {
  const c = create(<Home />);
  expect(c.toJSON()).toMatchSnapshot();
});
