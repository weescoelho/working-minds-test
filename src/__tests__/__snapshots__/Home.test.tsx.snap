// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`render Home page 1`] = `
Array [
  <header
    className="sc-dlnjwi kgUbIl"
  >
    <div
      className="sc-gtsrHT mSBoN"
    >
      CRUD
    </div>
  </header>,
  <div
    className="sc-eCApnc fCrHnU"
  >
    <div
      className="sc-jSFjdj bpKePK"
    />
    <button
      className="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedPrimary"
      disabled={false}
      onBlur={[Function]}
      onClick={[Function]}
      onDragLeave={[Function]}
      onFocus={[Function]}
      onKeyDown={[Function]}
      onKeyUp={[Function]}
      onMouseDown={[Function]}
      onMouseLeave={[Function]}
      onMouseUp={[Function]}
      onTouchEnd={[Function]}
      onTouchMove={[Function]}
      onTouchStart={[Function]}
      tabIndex={0}
      type="button"
    >
      <span
        className="MuiButton-label"
      >
        Adicionar estado
      </span>
    </button>
  </div>,
]
`;
